Added Webhooks support for docket alerts

albertisfu opened the pull request:
The Webhook page was updated according to the Figma proposal, some screenshots are below:

**Webhooks page, no webhooks registered.**
![Screen Shot 2022-07-11 at 10 01 06](https://user-images.githubusercontent.com/486004/178296500-9e59054d-04e1-49c2-99a3-140e630cbe9c.png)

**Add Webhook modal:**
According to the requirements.  Right now it's possible for a user to have only one webhook for each type.
When adding a new Webhook, the Event Type field will only show the users' remaining event types options.
![Screen Shot 2022-07-11 at 10 02 36](https://user-images.githubusercontent.com/486004/178296691-6a826b4a-4c10-413d-90f2-4cb190129889.png)

When a user tries to add a new Webhook after adding one webhook of each type, a message will be shown.
![Screen Shot 2022-07-11 at 10 04 46](https://user-images.githubusercontent.com/486004/178297643-d1faeb79-50ce-4ab5-b547-86f2408b3af6.png)

When updating a Webhook it's only possible to change the URL and enabled status. To avoid conflicts due to duplicating event types.
![Screen Shot 2022-07-11 at 10 03 40](https://user-images.githubusercontent.com/486004/178298195-c056f844-9d91-432d-91b9-9d0fe11f62b1.png)

When deleting a Webhook a confirmation dialog is shown:
![Screen Shot 2022-07-11 at 10 05 39](https://user-images.githubusercontent.com/486004/178299117-5b6f4e00-313e-4c65-b707-f28ee6aa6591.png)

Some additional comments:

- As described above, the one type webhook constraint for each user works at the interface level.  Should we also add a DB constraint? I didn't add it due to the comment that it might be needed to add support for multiple webhooks of each type in the future.

- According to requirements, Webhooks for DocketAlerts and Recap.email are now the same. Some fixes and changes were applied to accomplish it. The send_docket_to_webhook task was moved into alerts/tasks to avoid a circular reference.

- Webhooks for DocketAlerts are sent to users that have an enabled DOCKET_ALERT Webhook. The webhook is triggered only for cases to which the user is subscribed. For recap.email subscriptions, the webhook is triggered the first time only if the user has the "auto-subscribe" option enabled, took advantage of the existing logic to avoid doing additional queries to obtain webhook recipients.

- One question that popped up, is it necessary to retry the `send_docket_to_webhook` task on a RequestException or Timeout? Since we are going to count the webhook failures.

- I'll add support for Search Alerts and Recap Fetch webhooks in the next PRs.

Please let me know if additional changes are needed over the interface or logic.

mlissner commented on cl/users/urls.py:
```diff
@@ -102,6 +102,21 @@
     path("profile/api-token/", views.view_api_token, name="view_api_token"),
     path("profile/api-usage/", views.view_api_usage, name="view_api_usage"),
     path("profile/webhooks/", views.view_webhooks, name="view_webhooks"),
+    path(
+        "profile/webhook-form/",
+        views.webhook_form,
+        name="webhook_form",
+    ),
+    path(
+        "profile/webhooks-list/",
+        views.view_webhooks_list,
+        name="view_webhooks_list",
+    ),
+    path(
+        "profile/webhook-delete/<int:pk>/",
+        views.delete_webhook,
+        name="delete_webhook",
```
I just read many tutorials on HTMX and Django, and I don't like any of them. They all take the approach you did here, which uses multiple views separating out logic. Doing this leaves our views untethered to each other and spreads our code out very thinly. 

I think a better way to do this is to do as few views as possible, and to use HTTP verbs. So:

 - HTTP DELETE sent to /api/rest/v3/webhooks/<id>/ deletes something
 - HTTP PUT sent to /api/rest/v3/webhooks/<id>/ updates something
 - HTTP POST sent to /api/rest/v3/webhooks/ creates something
 - HTTP GET sent to /api/rest/v3/webhooks/ lists them
 - HTTP GET sent to /api/rest/v3/webhooks/<id> lists one (if needed)

This mostly mirrors the approach in Django REST Framework, and you can see that I used its URLs in my above concept. 

I spent some time looking for an integration between DRF and HTMX, but didn't find anything good. It seems like it should exist, but I guess not (yet). 

In any case, I think we should do something like the above. If we can use DRF to power HTMX, we should do that, using the URLs above. If we can't, we should use HTTP verbs and as few URL paths as possible, probably putting them at `/webhooks/` instead of under `/api/rest/v3/`.

Does this make sense? Do you agree?

mlissner replied:
The other thing that might be worth doing is to use django-htmx to combine views. For example, you have two that basically just list webhooks. Those could be combined by using the [django-htmx middleware](https://django-htmx.readthedocs.io/en/latest/middleware.html#django_htmx.middleware.HtmxDetails.__bool__).

albertisfu replied:
Thanks! this is a better approach to implementing htmx with django, I managed to use DRF to render over an HTML template (needed for htmx) using a ModelViewSet I could implement all the necessary actions for creating, editing, updating, and deleting webhooks.

Now it's possible to handle all the actions using one route: `/api/rest/v3/webhooks` and different verbs such as DELETE, PUT, POST, and GET.

In order to use `/api/rest/v3/webhooks` added this router on `cl.api.urls `, however, there is something that makes me think maybe that's wrong.  Because if you go to `/api/rest/v3/` it's possible to see the webhooks endpoint but this is not a normal API endpoint due to its actions render to an HTML template as default, would it be better to move it out of `/api/rest/v3/`?

mlissner replied:
Yes, I agree, now that I see it in /api/rest/v3/webhooks, it feels wrong. Let's give it a better home. A few possibilities:

 - /api/htmx/webhooks/  - I kind of like this.
 - /profile/webhooks/ - Feels limited, but maybe it's fine?
 - /webhooks/ - This feels like it's too high in the hierarchy.

What do you think?

mlissner replied:
I think this URL tweak is still needed.

albertisfu replied:
Sorry, I missed this comment. I've changed it to: `/api/htmx/webhooks/ ` I think it describes pretty well that it belongs to htmx actions.

I moved it into users.urls

mlissner commented on cl/users/views.py:
```diff
@@ -712,20 +713,69 @@ def moosend_webhook(request: HttpRequest) -> HttpResponse:
+def webhook_form(request: HttpRequest) -> HttpResponse:
+    """Handle the creation or editing of a webhook"""
+    pk = ""
+    if "pk" in request.GET:
+        pk = request.GET["pk"]
+    if "pk" in request.POST:
+        pk = request.POST["pk"]
```
Maybe this comment won't matter if we redo this, but why do we allow a GET request here?

mlissner commented on cl/users/templates/includes/webhooks-modal.html:
```diff
@@ -0,0 +1,20 @@
+      <div class="modal-header">
+        <button type="button" class="close" data-dismiss="modal" aria-label="Close"><span aria-hidden="true">&times;</span>
+        </button>
+        <h2 class="modal-title">Add a Webhook endpoint</h2>
```
You capitalized webhook a lot in here. Let's keep them uncapitalized most of the time:

```suggestion
        <h2 class="modal-title">Add a webhook endpoint</h2>
```

albertisfu replied:
changed!

mlissner commented on cl/users/templates/profile/webhooks.html:
```diff
@@ -1,68 +1,96 @@
+{% block footer-scripts %}
+  {% if DEBUG %}
+    <script src="{% static "js/htmx.js" %}"></script>
+    <link href="{% static "css/bootstrap4-toggle.css" %}" rel="stylesheet">
```
Did you end up using this?

albertisfu replied:
no, we don't use the toggle here, removed.

mlissner commented on cl/users/templates/profile/webhooks.html:
```diff
@@ -1,68 +1,96 @@
+        // If user has reached the max number of webhooks, show a message.
+        let event_type_options = document.getElementById("id_event_type").options.length;
+        if(event_type_options == 0){
+          webhook_form.innerHTML = "<b class='text-center'>You've reached the maximum number of webhooks. Please delete one before adding another.</b>";
```
```suggestion
          webhook_form.innerHTML = "<b class='text-center'>You already have a webhook configured for each type of event. Please delete one before making another.</b>";
```

albertisfu replied:
changed!

mlissner commented on cl/users/templates/profile/webhooks.html:
```diff
@@ -1,68 +1,96 @@
         </div>
       </div>
+      {% include "includes/webhooks-modal.html" %}
+      <p>Webhooks let you send a signal to your server whenever certain events happen on our website. To create a webhook, use the form below.</p>
```
```suggestion
      <p>Webhooks enable bi-directional APIs by sending data to your server whenever events happen on ours. Use webhooks to get the fastest updates &mdash; without polling.</p>
```

albertisfu replied:
changed!

mlissner commented on cl/users/templates/includes/webhooks-list.html:
```diff
@@ -0,0 +1,40 @@
+    <td>{{ webhook.url }}</td>
+    <td class="right" hx-headers='{"X-CSRFToken": "{{ csrf_token }}"}'>
+      <button hx-post="{% url "delete_webhook" webhook.pk %}"
+              hx-confirm="Are you sure you want to delete this Webhook?"
```
```suggestion
              hx-confirm="Are you sure you want to delete this webhook?"
```

albertisfu replied:
changed.

mlissner commented on cl/recap/tasks.py:
```diff
@@ -1658,11 +1606,8 @@ def process_recap_email(
-                    docket.pk, start_time, recap_email_recipients
-                ),
-                send_docket_to_webhook.si(docket.pk, start_time, epq.pk),
+            send_docket_alert.si(
+                docket.pk, start_time, recap_email_recipients
```
If we're not doing a chain and don't need special stuff provided by apply_async, we should just use .delay().

albertisfu replied:
changed!

mlissner replied:
There's an even better way to do this. Check out the example here, which drops the `si` too: https://docs.celeryq.dev/en/stable/userguide/calling.html#basics

albertisfu replied:
Thanks, I've changed it.

mlissner commented on cl/alerts/tasks.py:
```diff
@@ -163,7 +222,7 @@ def send_docket_alert(
 
     # Notify every user that's subscribed to this alert.
-    messages = make_alert_messages(
+    messages, da_webhook_recipients = make_alert_messages(
```
We need refactoring here. This is the general stuff we do:

 - Get recipients for messages
 - Get recipients for webhooks
 - Generate emails
 - Generate webhooks
 - Send emails
 - Send webhooks

Right now we call send_docket_alerts, which calls make_alert_messages and send_docket_to_webhook. make_alert_messages, makes the messages, and gathers a list of webhook recipients, and sends messages. send_docket_to_webhook takes recipients from make_alert_messages, then generates webhooks and sends them. This is all very confusing! 

Can you see if you can simplify this so that, perhaps, it more closely mirrors the six steps above? Somebody calls something like send_alerts_and_webhooks, and then it gets recipients for each, generates content for each, and sends each?

albertisfu replied:
I've refactored these methods as suggested, I hope with these changes everything is more clear.

- Now somebody calls `send_alert_and_webhook` or `send_alerts_and_webhooks` to start the sending docket alerts and webhooks process.
- `send_alert_and_webhook` first calls `get_docket_alert_recipients` to get recipients for docket alerts.
- `get_docket_alert_recipients` returns recipients for each type of notification including the new admin's notification to inform about @recap.email users that don't exist.
- Then we call `send_recap_email_user_not_found` to send the notification to admins if there are any @recap.email users that don't exist.
- Then we compose each type of docket alert message:
  - `docket_alert_messages`: messages for users that are currently subscribed to the case
  - `re_first_time_auto_subscribe_messages`: messages for @recap.email users that receive a notification for the case and have the auto-subscribe option enabled.
  - `re_first_time_no_auto_subscribe_messages`: messages for @recap.email users that receive a notification for the case and have the auto-subscribe option disabled.
- Then we join the different types of messages and sent them at once.
- Finally, we get webhooks recipients and sent them.

mlissner replied:
Refactor looks good. I still wonder if it could be simplified more, but I'm happy with it. I think it's a big improvement. Hopefully it still works properly. That's a risk, but it's worth taking when the code gets so complex. Thanks for sticking with it through two rounds of review.

mlissner reviewed (changes requested):
Thanks for this, Alberto. I think this looks great and it works well, but I want some tweaks to the code that'll make our lives easier and better in the future. The two big ones are:

1. I think we should be using more restful patterns with HTMX. I'm surprised nobody is, but I wrote more about how I think it should work in a comment.

2. I think we've reached the point where the alert and webhook code has become hard to follow and maintain, and it needs a refactor. Details in a separate comment.

On top of those big ones, a few small ones:

1. We need `noop` on the migration.
4. The enabled box in the modal should be aligned with the text to the left. Bootstrap usually makes this easy (or I saw you wanted to use a toggle for this instead, but I think the toggle is probably the wrong design here).
5. We need to write webhook documentation for that link. Maybe do this as a follow-up issue? Do we have an issue for that? Do you want to make one?
6. We're generating webhook event logs, but not showing them. Do you have something in mind for how to show them? 

Thank you.

albertisfu commented:
Thank you for your comments and suggestions!

I still continue working on the refactor of send docket alerts and webhooks methods but I wanted to submit these changes and comments to hear your opinion.

- I've applied the refactor of webhook views to use a Restful approach with HTMX. It was possible to use DRF for it, there is a comment about this above.
- I've opened the issue to write webhook documentation here #2176
- About Logs, yeah I think we could add a section to show logs to users, I added a proposal on [Figma](https://www.figma.com/file/l3hda5VzQw6QZyKckqlne9/Webhooks?node-id=0%3A1) for it with some comments, some screenshots below.
- I also was thinking that it would be helpful for users to add a feature that users can use to test its webhook integration. It might be something simple as allowing users to trigger an example webhook request for each event type, what do you think?
- I'll made some research about webhooks retries to make a proposal of retry back off, I have a couple of ideas on mind but I just want to be sure how other services handle webhooks retries.



![Screen Shot 2022-07-15 at 12 47 26](https://user-images.githubusercontent.com/486004/179281888-37164319-1000-4cb2-a70f-f6ed014a0c8e.png)
![Screen Shot 2022-07-15 at 12 47 41](https://user-images.githubusercontent.com/486004/179281928-d9fdf301-0d87-4d34-a145-69eda90c9f4e.png)

albertisfu commented:
I've submitted additional changes that refactor send docket alerts and webhooks tasks. I hope now it's simpler to follow the process, more details about this refactoring are in a comment above.

Let me know what you think.

Thanks

mlissner commented:
> I also was thinking that it would be helpful for users to add a feature that users can use to test its webhook integration. It might be something simple as allowing users to trigger an example webhook request for each event type, what do you think?

Yes, we should make an issue for this and put it on our backlog.

mlissner commented on cl/alerts/tasks.py:
```diff
@@ -28,117 +30,142 @@ def enqueue_docket_alert(d_pk: int) -> bool:
+def send_alert_and_webhook(
     d_pk: int,
     since: datetime,
     recap_email_recipients: list[str] = [],
```
I guess I missed this before, but `[]` is mutable and will cause all kinds of fun problems, probably. Instead this should be `= None`, and then you should add a something like:

```
if recap_email_recipients is None:
    recap_email_recipients = []
```

albertisfu replied:
Done, I've chanted it to `= None` and applied the suggestion:
```
if recap_email_recipients is None:
    recap_email_recipients = []
```
Thanks.

mlissner commented on cl/alerts/tasks.py:
```diff
@@ -28,117 +30,142 @@ def enqueue_docket_alert(d_pk: int) -> bool:
+    d_pk: int,
+    recap_email_recipients: list[str] = [],
+) -> tuple[
+    list[tuple[str, str]],
+    list[tuple[str, str]],
+    list[tuple[str, str]],
+    list[str],
```
OK, this refactor isn't quite it, and this return type should have been the "[code-smell](https://en.wikipedia.org/wiki/Code_smell)" that made you realize something was wrong. 

Linus Torvalds once said something like, "when you've got your data objects sorted out and as simple as possible, you know you're done." This ain't that. 

The solution here is hinted at by the name of the function. We want to get recipients, so it's going to be a list. But a list of what? We could use dicts like:

```
email: 'mike@free.law',
secret_key: 'asdfasdfasdfa',
first_time: True,
auto_subscribe: False,
does_not_exist: False,
```

Or, dare I say it: We could use an object to hold that information (this would make typing easier).

Then the return from this function would be something like `list(MessageData)` (or whatever). This is all notional, meaning I want you to make it good, but this return type is a major code smell.

mlissner replied:
Oh, I guess a data class is probably a good tool here too: https://docs.python.org/3/library/dataclasses.html

albertisfu replied:
I imagined I should have to use something like this, but I wasn't sure what kind of data structure was appropriate so I ended up using that ugly list of tuples. The new data class `DocketAlertRecipient` helps simplify all the things here and is clearer.

I just keep the `recap_email_user_does_not_exist_list` because is just a list of emails so it doesn't match with the new data class.

Thank you!

mlissner replied:
Looks good. Thanks again.

mlissner commented on cl/alerts/tasks.py:
```diff
@@ -28,117 +30,142 @@ def enqueue_docket_alert(d_pk: int) -> bool:
+        docket_id=d_pk, alert_type=DocketAlert.SUBSCRIPTION
+    )
+    for da in docket_alerts_current_subscribers:
+        recipient_tuple = (da.user.email, da.secret_key)
```
We need to remember to focus on queries in this code. So if you need to get `da.user.email`, remember to use a `select_related` up above.

albertisfu replied:
Perfect, I added it here and in a different query below where it's also needed to access a related object field.

mlissner commented on cl/alerts/tasks.py:
```diff
@@ -28,117 +30,142 @@ def enqueue_docket_alert(d_pk: int) -> bool:
+            # If a docket alert exists for this @recap.email user, avoid
+            # sending the first email
+            continue
+        if user_profile.auto_subscribe:
+            # First time recipient, auto_subscribe True
+            docket_alert = DocketAlert.objects.create(
+                docket_id=d_pk, user=user_profile.user
+            )
+            recipient_tuple = (email_address, docket_alert.secret_key)
+            recap_email_first_time_auto_subscribe_list.append(recipient_tuple)
+        else:
+            # First time recipient, auto_subscribe False
+            docket_alert = DocketAlert.objects.create(
+                docket_id=d_pk,
+                user=user_profile.user,
+                alert_type=DocketAlert.UNSUBSCRIPTION,
+            )
+            recipient_tuple = (email_address, docket_alert.secret_key)
+            recap_email_first_time_no_auto_subscribe_list.append(
```
This `objects.create(` code is almost completely duplicated. I think when you make the refactor to return a list of dicts/objects, this should simplify. It should wind up with something like:

```
    # Somewhere above: md = MessageData()
    alert_type = DocketAlert.SUBSCRIPTION if auto_subscribe else DocketAlert.UNSUBSCRIPTION
    docket_alert = DocketAlert.objects.create(
        docket_id=d_pk,
        user=user_profile.user,
        alert_type=alert_type,
    )
    md.email_address = email_address
    md.secret_key = docket_alert.secret_key
    md.auto_subscribe = user_profile.auto_subscribe
    message.append(md)
return messages
```

albertisfu replied:
Yeah, this simplification was possible thanks to the new data class.

mlissner commented on cl/alerts/tasks.py:
```diff
@@ -168,25 +200,58 @@ def send_docket_alert(
-    # Notify every user that's subscribed to this alert.
-    messages = make_alert_messages(
-        d, new_des, email_addresses, recap_email_recipients
+    docket_alert_messages = make_alert_messages(
+        d,
+        new_des,
+        docket_alert_recipients,
+        first_email=False,
+        auto_subscribe=False,
+    )
+    re_first_time_auto_subscribe_messages = make_alert_messages(
+        d,
+        new_des,
+        recap_email_auto_subscribe_recipients,
+        first_email=True,
+        auto_subscribe=True,
+    )
+    re_first_time_no_auto_subscribe_messages = make_alert_messages(
+        d,
+        new_des,
+        recap_email_no_auto_recipients,
+        first_email=True,
+        auto_subscribe=False,
+    )
+
+    messages = (
+        docket_alert_messages
+        + re_first_time_auto_subscribe_messages
+        + re_first_time_no_auto_subscribe_messages
     )
```
After the refactor I mention above, I think this will all drop away. Aaaaaahhhh.

albertisfu replied:
Yeah, this was removed!

mlissner commented on cl/alerts/tasks.py:
```diff
@@ -28,117 +30,142 @@ def enqueue_docket_alert(d_pk: int) -> bool:
     subject_template = loader.get_template("docket_alert_subject.txt")
     subject_context = {
         "docket": d,
         "count": new_des.count(),
```
You'll save a query by making this `len(new_des)`. Just saying!

albertisfu replied:
Changed!

mlissner commented on cl/alerts/tasks.py:
```diff
@@ -244,3 +309,76 @@ def send_unsubscription_confirmation(
+
+    :param d_pk: The Docket primary key
+    :param since: Start time for querying the docket entries
+    :webhook_recipients: The list of email recipients to send the webhook to
```
```suggestion
    :param webhook_recipients: The list of email recipients to send the webhook to
```

mlissner commented on cl/alerts/tasks.py:
```diff
@@ -168,25 +200,58 @@ def send_docket_alert(
+    )
+    # Send the docket to webhook
+    webhook_email_recipients = [email[0] for email in webhook_recipients]
+    send_docket_to_webhook.delay(d_pk, since, webhook_email_recipients)
```
Passing the email addresses down to the webhook is an interesting optimization, but I think it's not worth the confusion ("Why are we sending email addresses to the webhook?") or the tethering ("Why do we need to send alerts before we can send webhooks?").

Also, the name of the function is a bit odd. Shouldn't it be `send_docket_alert_webhooks`?

So, with those two changes, I think the signature here should change to:
```
def send_docket_alert_webhooks(d_pk: int, since: datetime):
```

albertisfu replied:
Changed, now it's not necessary to pass the recipients to `send_docket_alert_webhooks`.

About the question:
`Why do we need to send alerts before we can send webhooks?`

We don't need to send alerts before we send webhooks but we need to get the recipients of docket alerts before sending the webhooks because when getting the recipients we also create docket alerts for recap email users with auto-subscribe turned on, so we also include those docket alert subscriptions (first-time notifications for a case) when sending webhooks.

mlissner commented on cl/alerts/tasks.py:
```diff
@@ -244,3 +309,76 @@ def send_unsubscription_confirmation(
+        # No new docket entries.
+        return
+
+    webhooks = Webhook.objects.filter(
+        event_type=WebhookEventType.DOCKET_ALERT,
+        user__email__in=webhook_recipients,
+        enabled=True,
```
In my last comment I said not to pass down the webhook_recipients variable, so now we need to tweak this query to get "Enabled webhooks of type DOCKET_ALERT where the user is subscribed to d_pk."

I admit I gave this a bit of thought and couldn't get it quite figured out in my head, but I'm sure you'll nail it.

albertisfu replied:
Yeah, it was possible to get the recipients within `send_docket_alert_webhooks`

First, get a list of recipients subscribed to the docket.
```
webhook_recipients = User.objects.filter(
            docket_alerts__docket_id=d_pk,
            docket_alerts__alert_type=DocketAlert.SUBSCRIPTION,
    ).distinct()```
    
Then select the DocketAlert Webhooks from these users.
```
webhooks = Webhook.objects.filter(
        event_type=WebhookEventType.DOCKET_ALERT,
        user__in=webhook_recipients,
        enabled=True,
    )
```

mlissner replied:
I still think it's possible to do it in one query, but I won't ask you to try. Two is great (and actually pretty clear). :)

mlissner commented on cl/alerts/tasks.py:
```diff
@@ -244,3 +309,76 @@ def send_unsubscription_confirmation(
+            "webhook": {
+                "event_type": webhook.event_type,
+                "version": webhook.version,
+                "date_created": webhook.date_created.isoformat(),
```
Don't forget to do the `deprecation_date` field (or similar) here.

albertisfu replied:
I've added it. Now, is set `None`.

mlissner commented on cl/alerts/tasks.py:
```diff
@@ -244,3 +309,76 @@ def send_unsubscription_confirmation(
+            response=response.text,
+        )
+
+        if not response.ok:
+            webhook.failure_count = webhook.failure_count + 1
+            webhook.save()
```
We can avoid this query by checking response.ok before the `create`, right?

albertisfu replied:
Well, this `failure_count` belongs to the Webhook object, not the WebhookEvent. That's why is not passed to the `create`.

mlissner replied:
Got it.

mlissner commented on cl/alerts/tasks.py:
```diff
@@ -244,3 +309,76 @@ def send_unsubscription_confirmation(
+
+    serialized_docket_entries = []
+    for de in docket_entries:
+        serialized_docket_entries.append(DocketEntrySerializer(de).data)
```
I'm a bit surprised you can get away with such a simple serialization. Check out the hoops we have to go through for the bulk data serializer. https://github.com/freelawproject/courtlistener/blob/0c74a05a74188066f268c7023f2a12b0f583cb8e/cl/api/tasks.py#L160-L166

albertisfu replied:
I think I forget to comment on this:
Interesting. Yeah, I think using the `JSONRenderer()` we ensure the JSON we sent to Webhooks meets the same characteristics we have in API responses. So I changed to use `JSONRenderer()`. Before we were using the JSON encoding included on `requests` where you just needed to pass a `dict`.

mlissner reviewed (changes requested):
A lot of comments and suggestions, sorry. But my review is finally done. Two more little ones for HTMX:

1. Does it make sense to put HTMX templates into a subdirectory of `includes`? 
2. No tests for HTMX? I'm surprised. :)

mlissner commented:
The figma design for logs looks great. No further comments, but let's get an issue going for that if we don't have one already, please. Thank you!

albertisfu commented:
Thank you for all your suggestions, they were so useful. 
I hope that with this refactor the code is simpler.

In addition to some comments above. I added the `htmx` templates within subdirectory of `includes`.

Added tests for HTMX endpoints.

I've opened an issue for logs and another for the webhook test feature for users.

mlissner commented on cl/alerts/tasks.py:
```diff
@@ -168,25 +195,26 @@ def send_docket_alert(
-    messages = make_alert_messages(
-        d, new_des, email_addresses, recap_email_recipients
-    )
+    docket_alert_messages = make_alert_messages(d, new_des, da_recipients)
+    messages = docket_alert_messages
```
What's this about?

albertisfu replied:
Oh yeah, that's not needed, I've removed it.

mlissner commented:
One or two things left, but easy stuff. Back to you, thank you.

albertisfu commented:
I've applied these small changes.
Let me know if there is something else.
Thanks again!
